### PR TITLE
add xmmpost() to back end

### PR DESCRIPTION
**Pull request by WalterBright**
Adds SIMD code generation to x++ and x-- operations.

### CODE REVIEW

**Comment by ibuclaw** [comment]
Doesn't the frontend need to allow post increments on simd data also?  Or is that already in place or to be added later.

**Comment by WalterBright** [comment]
I don't know what you mean. The code was previously using the x87 for postincrements.

**Comment by ibuclaw** [comment]
Ah ok, so this is clearing up old x87 clutter then.

File: src/backend/cod2.c [hunk 1]
```diff
@@ -4609,11 +4609,14 @@ code *cdpost(elem *e,regm_t *pretregs)
...
-#else
+        if (config.fpxmmregs && tyxmmreg(tyml) &&
+            !tycomplex(tyml) // SIMD code is not set up to deal with complex
+           )
+            return xmmpost(e,pretregs);
+
         if (config.inline8087)
-                return post87(e,pretregs);
+            return post87(e,pretregs);
+#if WINDOS
```
Comment (yebblies) [inline src/backend/cod2.c]: Doesn't look right...

Comment (WalterBright) [inline src/backend/cod2.c]: fixed

File: src/backend/cgxmm.c [hunk 2]
```diff
@@ -476,6 +476,92 @@ code *xmmopass(elem *e,regm_t *pretregs)
     return cat4(cr,cl,cg,co);
 }
 
+/********************************
+ * Generate code for post increment and post decrement.
+ */
+
+code *xmmpost(elem *e,regm_t *pretregs)
```
Comment (andralex) [inline src/backend/cgxmm.c]: I see this codebase doesn't have space after comma...

File: src/backend/cgxmm.c [hunk 3]
```diff
@@ -476,6 +476,92 @@ code *xmmopass(elem *e,regm_t *pretregs)
...
+    if (config.flags4 & CFG4optimized)
+    {
+        // Be careful of cases like (x = x+x+x). We cannot evaluate in
+        // x if x is in a register.
+        unsigned varreg;
+        regm_t varregm;
+        if (isregvar(e1,&varregm,&varreg) &&    // if lvalue is register variable
+            doinreg(e1->EV.sp.Vsym,e2)          // and we can compute directly into it
+           )
+        {   regvar = TRUE;
```
Comment (andralex) [inline src/backend/cgxmm.c]: We really should dispense with "code after brace" in new code.

**Review by andralex — APPROVED** [review]

File: src/backend/cgxmm.c (hunk 2 above)
Comment (WalterBright) [inline src/backend/cgxmm.c]: The back end does not have that style.

File: src/backend/cgxmm.c (hunk 3 above)
Comment (WalterBright) [inline src/backend/cgxmm.c]: I know, much of that is the legacy of 24 line displays. (This particular piece was cut/pasted from elsewhere.)

**Comment by WalterBright** [comment]
Auto-merge toggled on

**Pull request merged**